backend/src/schema/resolvers/notifications.js

```diff
@@ -48,16 +48,20 @@ export default {
       const readTxResultPromise = session.readTransaction(async transaction => {
         const notificationsTransactionResponse = await transaction.run(
           ` 
-          MATCH (resource {deleted: false, disabled: false})-[notification:NOTIFIED]->(user:User {id:$id})
-          ${whereClause}
-          WITH user, notification, resource,
-          [(resource)<-[:WROTE]-(author:User) | author {.*}] as authors,
-          [(resource)-[:COMMENTS]->(post:Post)<-[:WROTE]-(author:User) | post{.*, author: properties(author)} ] as posts
-          WITH resource, user, notification, authors, posts,
-          resource {.*, __typename: labels(resource)[0], author: authors[0], post: posts[0]} as finalResource
-          RETURN notification {.*, from: finalResource, to: properties(user)}
-          ${orderByClause}
-          ${offset} ${limit}
+            MATCH (resource)-[notification:NOTIFIED]->(user:User {id:$id})
+            WHERE
+              ((labels(resource)[0] in ["Post", "Comment"] AND NOT resource.deleted AND NOT resource.disabled)
+              OR labels(resource)[0] in ["Report"])
+              ${whereClause}
+            WITH user, notification, resource,
+            [(resource)<-[:WROTE]-(author:User) | author {.*}] AS authors,
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-23T18:52:44Z" title="Thursday, January 23rd 2020, 7:52:44 pm +01:00">Jan 23, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -48,16 +48,18 @@ export default {
       const readTxResultPromise = session.readTransaction(async transaction => {
         const notificationsTransactionResponse = await transaction.run(
           ` 
-          MATCH (resource {deleted: false, disabled: false})-[notification:NOTIFIED]->(user:User {id:$id})
-          ${whereClause}
-          WITH user, notification, resource,
-          [(resource)<-[:WROTE]-(author:User) | author {.*}] as authors,
-          [(resource)-[:COMMENTS]->(post:Post)<-[:WROTE]-(author:User) | post{.*, author: properties(author)} ] as posts
-          WITH resource, user, notification, authors, posts,
-          resource {.*, __typename: labels(resource)[0], author: authors[0], post: posts[0]} as finalResource
-          RETURN notification {.*, from: finalResource, to: properties(user)}
-          ${orderByClause}
-          ${offset} ${limit}
+            MATCH (resource)-[notification:NOTIFIED]->(user:User {id:$id})
+            WHERE ((labels(resource)[0] in ["Post", "Comment"] AND NOT resource.deleted AND NOT resource.disabled) OR labels(resource)[0] in ["Report"])
+              ${whereClause}
+            WITH user, notification, resource,
+            [(resource)<-[:WROTE]-(author:User) | author {.*}] as authors,
+            [(resource)-[:COMMENTS]->(post:Post)<-[:WROTE]-(author:User) | post {.*, author: properties(author)} ] as posts,
+            [(reportedResource)<-[:BELONGS_TO]-(resource)<-[file:FILED]-(user) | file {.*, reportedResource: apoc.map.merge(properties(reportedResource), {__typename: labels(reportedResource)[0]})} ] as files
```


```suggestion
            [(reportedResource)<-[:BELONGS_TO]-(resource:Report)<-[file:FILED]-(user) | file {.*, reportedResource: apoc.map.merge(properties(reportedResource), {__typename: labels(reportedResource)[0]})} ] as files
```
I think it probably doesn't matter cause at the moment you can only `file` a `Report`, but it shouldn't hurt either, and if we do add other possible edges like a `User` `FILED` a `COMPLAINT`, or something else, it might need to be refactored..
Or I might have missed something. What do you think?

I have never seen `apoc.map.merge(properties(reportedResource)` before, is it black magic?? Seems cool, is there some test I can look at, that shows what this does?

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2020-01-24T09:43:04Z" title="Friday, January 24th 2020, 10:43:04 am +01:00">Jan 24, 2020</time>_

---


It was the only way to get the `__typename` property into the `reportedResource`, I found.
It was described somewhere in the Apoc docs.
A `map` here is like a JavaScript object. Here you have two maps (objects) with properties and merge them together.

Like:
```
{ ...properties(reportedResource),  ...{__typename: labels(reportedResource)[0]}}
```
Shorter:
```
{ ...reportedResource,  __typename: labels(reportedResource)[0]}
```